src/integrate/scarches/script.py

```diff
@@ -0,0 +1,179 @@
+    """
+    model = _detect_base_model(model_path)
+    model.prepare_query_anndata(adata_query, model_path)
+
+    # Load query data into the model
+    vae_query = model.load_query_data(
+            adata_query,
+            model_path,
+            freeze_dropout=True
+    )
+
+    # Train scArches model for query mapping
+    vae_query.train(
+            max_epochs=par["max_epochs"],
+            early_stopping=par['early_stopping'],
+            early_stopping_monitor=par['early_stopping_monitor'],
+            early_stopping_patience=par['early_stopping_patience'],
+            early_stopping_min_delta=par['early_stopping_min_delta'],
+            check_val_every_n_epoch=check_val_every_n_epoch,
+            use_gpu=(cuda_is_available() or mps_is_available())
+    )
+
+    return vae_query, adata_query
+
+
+def _download_HLCA_reference_model(directory):
+    import os
+
+    HLCA_PATH = "https://zenodo.org/record/6337966/files/HLCA_reference_model.zip"
+
+    os.system(f"curl  {HLCA_PATH} --output {directory}/HLCA_reference_model.zip")
+    os.system(f"unzip {directory}/HLCA_reference_model.zip -d {directory}")
+
+    scvi.model.SCANVI.convert_legacy_save(f"{directory}/HLCA_reference_model", f"{directory}/HLCA_reference_model_new")
+
+    return f"{directory}/HLCA_reference_model_new"
+
+
+def main():
+    SUPPORTED_REFERENCES = ["HLCA"]
+
+    if par["reference"] is not None and par["reference"] not in SUPPORTED_REFERENCES:
+        raise ValueError(f"{par['reference']} is not supported reference. Please select one of {', '.join(SUPPORTED_REFERENCES)}")
+
+    logger = _setup_logger()
+
+    mdata_query = mudata.read(par["input"].strip())
+    adata_query = mdata_query.mod[par["modality"]]
+
+    if "dataset" not in adata_query.obs.columns:
+        # Write name of the dataset as batch variable
+        if par["dataset_name"] is None:
+            logger.info("Detecting dataset name")
+            par["dataset_name"] = extract_file_name(par["input"])
+            logger.info(f"Detected {par['dataset_name']}")
+
+        adata_query.obs["dataset"] = par["dataset_name"]
+
+    if par["reference"] == "HLCA":
+        import tempfile
+
+        with tempfile.TemporaryDirectory() as directory:
+            model_path = _download_HLCA_reference_model(directory)
+            vae_query, adata_query = map_to_existing_reference(adata_query, model_path=model_path)
+            model_name = _read_model_name_from_registry(model_path)
+        
+    else:
+        raise ValueError(f"Reference {par['reference']} is not supported")
+
+    # Save info about the used model
+    adata_query.uns["integration_method"] = model_name
+
+    logger.info("Trying to write latent representation")
+    output_key = par["obsm_output"].format(model_name=model_name)
+    adata_query.obsm[output_key] = vae_query.get_latent_representation()
+
+    logger.info("Reassigning modality")
+    mdata_query.mod[par["modality"]] = adata_query
+    try:
```

Comment:
Could you explain why this happens/is needed?

Reply:
I faced a weird error when trying to reassign modality. This is discussed here: https://github.com/scverse/mudata/issues/27

Briefly, after assigning `adata_query` to the modality of `mdata_query`, described shape of `mdata_query` might be inconsistent with a number of vars and shape of the data. When trying to save such object to a file, one can get a KeyError.

It can be solved by running `mdata_query.update()`. Weirdly, sometimes it still throws a KeyError, but then the shape of the mdata and the problem with saving it to the file is fixed.

Reply:
By the way, according to the comments of mudata developers, reassigning modalities is not expected way to work with mudata. But we use it quite often in this project

Reply:
This is not needed anymore, I decided to create new MuData object instead

Reply:
Good writeup! I learned a thing or two from the issue.